test/jdk/java/io/Reader/Of.java

```diff
@@ -0,0 +1,199 @@
+        assertEquals(new String(c),
+                CONTENT.substring(8, 16) + CONTENT.substring(0, 8),
+                "read(char[],int,int) provides wrong content");
+    }
+
+    @Test(dataProvider = "readers")
+    public void testReadBIILenZero(Reader reader) throws IOException {
+        assertEquals(reader.read(new char[1], 0, 0), 0,
+                "read(char[],int,int) != 0");
+    }
+
+    @Test(dataProvider = "readers")
+    public void testReadDirectCharBuffer(Reader reader) throws IOException {
+        CharBuffer charBuffer = ByteBuffer.allocateDirect(32).asCharBuffer();
+        charBuffer.position(8);
+        assertEquals(reader.read(charBuffer), 8,
+                "read(CharBuffer) does not respect position or limit");
+        charBuffer.rewind();
+        assertEquals(reader.read(charBuffer), 8,
+                "read(CharBuffer) does not respect end of stream");
+        charBuffer.rewind();
+        assertEquals(charBuffer.toString(),
+                // last part first proofs that copy loops correctly stopped
+                CONTENT.substring(8, 16) + CONTENT.substring(0, 8),
+                "read(CharBuffer) provides wrong content");
+    }
+
+    @Test(dataProvider = "readers")
+    public void testReadNonDirectCharBuffer(Reader reader) throws IOException {
+        CharBuffer charBuffer = CharBuffer.allocate(16);
+        charBuffer.position(8);
+        assertEquals(reader.read(charBuffer), 8,
+                "read(CharBuffer) does not respect position or limit");
+        charBuffer.rewind();
+        assertEquals(reader.read(charBuffer), 8,
+                "read(CharBuffer) does not respect end of stream");
+        charBuffer.rewind();
+        assertEquals(charBuffer.toString(),
+                CONTENT.substring(8, 16) + CONTENT.substring(0, 8),
+                "read(CharBuffer) provides wrong content");
+    }
+
+    @Test(dataProvider = "readers")
+    public void testReadCharBufferZeroRemaining(Reader reader) throws IOException {
+        CharBuffer charBuffer = CharBuffer.allocate(0);
+        assertEquals(reader.read(charBuffer), 0, "read(CharBuffer) != 0");
+    }
+
+    @Test(dataProvider = "readers")
+    public void testReady(Reader reader) throws IOException {
+        assertTrue(reader.ready());
+    }
+
+    @Test(dataProvider = "readers")
+    public void testSkip(Reader reader) throws IOException {
+        assertEquals(reader.skip(8), 8, "skip() does not respect limit");
+        assertEquals(reader.skip(9), 8, "skip() does not respect end of stream");
+        assertEquals(reader.skip(1), 0, "skip() does not respect empty stream");
+    }
+
+    @Test(dataProvider = "readers")
+    public void testTransferTo(Reader reader) throws IOException {
+        StringWriter sw = new StringWriter(16);
+        assertEquals(reader.transferTo(sw), 16, "transferTo() != 16");
+        assertEquals(reader.transferTo(sw), 0,
+                "transferTo() does not resect empty stream");
```

Comment:
Typo respect (I also wonder if “empty stream” (used multiple times) should be “end of stream” or “exhausted stream”. The test never get an empty stream to test.

Reply:
"empty stream" is actually correct. The case "end of stream" is implied in the line before (the test that expects result `16`, it starts with non-empty stream, it stops at end-of-stream). From that point, the stream is now empty, as `transferTo()` transferred *everything*. So what this test (the one expecting result `0`) tests is what happens if `transferTo()` starts in an already-empty situation.